### PR TITLE
SWC-6548 - Add utils and reducer for TableColumnSchemaEditor

### PR DESCRIPTION
**Pull request by nickgros**
This PR contains the bulk of the business logic for managing data in the Table Column Schema form in SWC-6548

### CODE REVIEW

File: packages/synapse-react-client/src/components/TableColumnSchemaEditor/TableColumnSchemaFormReducer.ts [hunk 1]
```diff
@@ -0,0 +1,500 @@
...
+    default:
+      throw new Error(`Unexpected action`, action)
+  }
+  return [...prevState]
+}
+
+export const tableColumnSchemaFormDataAtom = atomWithReducer<
+  ColumnModelFormData[],
+  TableColumnSchemaFormReducerAction
+>([], reducer)
```
Comment (nickgros) [inline packages/synapse-react-client/src/components/TableColumnSchemaEditor/TableColumnSchemaFormReducer.ts:500]: This `atomWithReducer` will be passed to the form React component and each change will be mediated by the reducerFn.

File: packages/synapse-react-client/src/components/TableColumnSchemaEditor/TableColumnSchemaFormReducer.ts [hunk 2]
```diff
@@ -0,0 +1,500 @@
...
+      toggleSelect(action, prevState)
+      break
+    }
+    case 'delete':
+      prevState = deleteColumnModel(prevState)
+      break
+    case 'move': {
+      prevState = moveColumnModel(action, prevState)
+      break
+    }
```
Comment (nickgros) [inline packages/synapse-react-client/src/components/TableColumnSchemaEditor/TableColumnSchemaFormReducer.ts:468]: `move` would handle Drag-N-Drop

File: packages/synapse-react-client/src/components/TableColumnSchemaEditor/TableColumnSchemaFormReducer.ts [hunk 3]
```diff
@@ -0,0 +1,500 @@
...
+      prevState = moveSelectedItemsDown(prevState, cm => cm.isSelected)
+      prevState.forEach(cm => {
+        if (cm.jsonSubColumns) {
+          cm.jsonSubColumns = moveSelectedItemsDown(
+            cm.jsonSubColumns,
+            jsc => jsc.isSelected,
+          )
+        }
+      })
+      break
```
Comment (nickgros) [inline packages/synapse-react-client/src/components/TableColumnSchemaEditor/TableColumnSchemaFormReducer.ts:490]: `moveUp`/`moveDown` would handle moving selected items (i.e. the current paradigm)

File: packages/synapse-react-client/src/components/TableColumnSchemaEditor/TableColumnSchemaFormReducer.ts [hunk 4]
```diff
@@ -0,0 +1,500 @@
...
+  action: {
+    type: 'setColumnModelValue'
+    columnModelIndex: number
+    jsonSubColumnModelIndex?: number
+    value: ColumnModelFormData | JsonSubColumnModelFormData
+  },
+  prevState: ColumnModelFormData[],
+) {
+  const { columnModelIndex, jsonSubColumnModelIndex, value } = action
+  if (
```
Comment (jay-hodgson) [inline packages/synapse-react-client/src/components/TableColumnSchemaEditor/TableColumnSchemaFormReducer.ts:300]: Should this be something like this (same comment for a subset of other utility functions)?
if (prevState && prevState[columnModelIndex]) {
  if (prevState[columnModelIndex].jsonSubColumns &&
    jsonSubColumnModelIndex !== undefined) { ...}
  else {...}
}

File: packages/synapse-react-client/src/components/TableColumnSchemaEditor/TableColumnSchemaFormReducer.ts (hunk 1 above)
Comment (jay-hodgson) [reply packages/synapse-react-client/src/components/TableColumnSchemaEditor/TableColumnSchemaFormReducer.ts:500]: Great pattern

**Review by jay-hodgson — APPROVED** [review]

File: packages/synapse-react-client/src/components/TableColumnSchemaEditor/TableColumnSchemaFormReducer.ts (hunk 4 above)
Comment (nickgros) [reply packages/synapse-react-client/src/components/TableColumnSchemaEditor/TableColumnSchemaFormReducer.ts:300]: Good idea, that would definitely be safer

**Pull request merged**